Allow importing realms during startup (#10754)
keycloak/keycloak: docs/guides/src/main/server/importExport.adoc

<#import "/templates/guide.adoc" as tmpl>
<#import "/templates/kc.adoc" as kc>

<@tmpl.guide
    title="Importing and Exporting Realms"
    summary="An overview about how to import and export realms">

In this guide, you are going to understand the different approaches for importing and exporting realms using JSON files.

== Exporting a Realm to a Directory

> To export a realm, you can use the `export` command.

Reviewer: ```suggestion
To export a realm, you can use the `export` command. Your Keycloak server instance must not be started when invoking this command.
```

Reviewer: Just to make sure people don't try to run the export command in e.g. another shell instance while running keycloak. Just tried it myself, then the export command fails (was expected from my side).

<@kc.export parameters="--help"/>

To export a realm to a directory, you can use the `--dir <dir>` option.

<@kc.export parameters="--dir <dir>"/>

When exporting realms to a directory, the server is going to create separate files for each realm being exported.

> 

Reviewer: think this is worth a subheading

You are also able to configure how users are going to be exported by setting the `--users <strategy>` option. The values available for this
option are:

> * *different_files*: Users export into different files subject to the maximum number of users per file. This is the default value for this option.

Author: Could you please also send a fix to the official documentation?

> * *different_files*: Users export into different files subject to the maximum number of users per file. This is the default value for this option.

Reviewer: "official documentation" - isn't that the guide you added to this PR? could you elaborate?

* *skip*: {project_name} skips exporting users.

> * *realm_file*:  Users export to the same file with the realm settings. The file is similar to "foo-realm.json" with realm data and users.

Author: Could you please also send a fix to the official documentation?

> * *realm_file*:  Users export to the same file with the realm settings. The file is similar to "foo-realm.json" with realm data and users.

Reviewer: see other comment

> * *same_file*:  Users export to the same file but different from the realm file. The result is similar to "foo-realm.json" with realm data and "foo-users.json" with users.

Reviewer: also a `-0` at end of filename, can we remove it to match the docs pls?

> * *same_file*:  Users export to the same file but different from the realm file. The result is similar to "foo-realm.json" with realm data and "foo-users.json" with users.

Author: Could you please also send a fix to the official documentation?

> * *same_file*:  Users export to the same file but different from the realm file. The result is similar to "foo-realm.json" with realm data and "foo-users.json" with users.

Reviewer: see other comment

If you are exporting users using the `different_files` strategy, you should also be able to set how many users per file you want by setting the `users-per-file` option. If not specified, the default value is `50`.

<@kc.export parameters="--dir <dir> --users different_files --users-per-file 100"/>

== Exporting a Realm to a File

To export a realm to a file, you can use the `--file <file>` option.

> <@kc.export parameters="--file <file>"/>

Reviewer: when i use e.g. `kc.sh export --file=realm.exe` this works and creates a theoretically executable file on my filesystem. can we add a convention that the filename to specify has to end with `json`?

Author: IMO, unnecessary.

Reviewer: not hard on this, so works for me

> When exporting realms to a file, the server is going to use the same file to store the configuration for all the realms being exported.

Reviewer: ```suggestion
When exporting realms to a file, keycloak will use the specified file to store all configuration for realms and users being exported.
```

Reviewer: users are exported directly when you use the file export. also e.g. `--users=skip` does not take effect, but also does not let the command fail, so is silently ignored. @stianst @pedroigor in terms of ux, should we fail fast when `--file <mypath>` `--users=...` is used instead of silently ignoring? I think yes, but not 100% sure.

Reviewer: another option would be to make user handling also available for the file command, but that also feels weird as its not creating one file then.

Author: Makes sense. Let us do it as a follow up.

<@kc.import parameters="--help"/>

After exporting a realm to a directory, you can use the `--dir <dir>` option to import the realm back to the server as follows:

<@kc.import parameters="--dir <dir>"/>

When importing realms using the `import` command, you are able to set if existing realms should be skipped, or if they should be overridden with the new configuration. For that,
you can set the `--override` option as follows:

<@kc.import parameters="--dir <dir> --override false"/>

By default, the `--override` option is set to `true` so that realms are always overridden with the new configuration.

== Importing a Realm from a File

To import a realm previously exported in a single file, you can use the `--file <file>` option as follows:

<@kc.import parameters="--file <file>"/>

== Importing a Realm during Startup

You are also able to import realms when the server is starting by using the `--import-realm` option.

<@kc.start parameters="--import-realm"/>

When you set the `--import-realm` option, the server is going to try to import any realm configuration file from the `data/import` directory. Each file in this directory should
contain a single realm configuration.

> If a realm already exists in the server, the import operation is skipped.

Reviewer: Feels a bit weird for me that we have different defaults for this one as for the import command. import command overwrites, here realms are skipped. Why is that? Should we align?

Author: We are re-enabling the legacy behavior as this is what people are expecting now in terms of feature parity.

We can probably use the same import strategy in both commands in the future.

Reviewer: works for me

Reviewer 2: I wouldn't necessarily expect that the auto-import works the same way like it did in legacy. At least, it will be possible again, no matter if I need to specify override/skip options additionally.

== Using Environment Variables within the Realm Configuration Files

> When importing a realm, you are able to use placeholders to resolve values from environment variables for any realm configuration.

Reviewer: love this one :)
